src/qibolab/instruments/qblox/cluster_qcm_bb.py

```diff
@@ -160,29 +146,24 @@ def connect(self, cluster: QbloxCluster = None):
         """
         if self.is_connected:
             return
-
-        elif cluster is not None:
-            self.device = cluster.modules[int(self.address.split(":")[1]) - 1]
-            # test connection with module
-            if not self.device.present():
-                raise ConnectionError(
-                    f"Module {self.device.name} not connected to cluster {cluster.name}"
-                )
-            # once connected, initialise the parameters of the device to the default values
-            self._device_num_sequencers = len(self.device.sequencers)
-            self._set_default_values()
-            # then set the value loaded from the runcard
-            try:
-                for port in self._ports:
-                    self._sequencers[port] = []
-                    self._ports[port].hardware_mod_en = True
-                    self._ports[port].nco_freq = 0
-                    self._ports[port].nco_phase_offs = 0
-            except Exception as error:
-                raise RuntimeError(
-                    f"Unable to initialize port parameters on module {self.name}: {error}"
-                )
-            self.is_connected = True
+        # test connection with module. self.device is initialized in QbloxController connect()
+        if not self.device.present():
+            raise ConnectionError(f"Module {self.device.name} not present")
+        # once connected, initialise the parameters of the device to the default values
+        self._device_num_sequencers = len(self.device.sequencers)
+        self._set_default_values()
+        # then set the value loaded from the runcard
+        try:
+            for port in self._ports:
+                self._sequencers[port] = []
+                self._ports[port].hardware_mod_en = True
+                self._ports[port].nco_freq = 0
+                self._ports[port].nco_phase_offs = 0
+        except Exception as error:
+            raise RuntimeError(
+                f"Unable to initialize port parameters on module {self.name}: {error}"
+            )
+        self.is_connected = True
```

Comment:
This function is quite repeated as is in the other modules.

I.e. the part that is written here (`QcmBb`) is almost completely repeated, while they may have extra bits inside the `try:` block.

I'd suggest lifting the `connect()` function to the base class, and leave to each subclass just the part inside the `try:` (just make an abstractmethod in the base class, `_setup_ports()` or something like, and implement in the subclasses).

Reply:
Thanks for the suggestion!

Reply:
Solved [0004e88](https://github.com/qiboteam/qibolab/pull/730/commits/0004e8855ec34d876d7303e6b53e4b55809f4654).